File: SortStack.py
def sort_stack(stack):
    # This temporary stack will hold elements in sorted order (smallest at bottom)
    tmp_stack = []
    
    while len(stack) > 0:
        # 1. Pop the top element from the original stack
        current = stack.pop()
        
        # 2. While tmp_stack is not empty and its top element is GREATER than current
        while len(tmp_stack) > 0 and tmp_stack[-1] > current:
            # Move elements back to the original stack to clear the way
            stack.append(tmp_stack.pop())
            
        # 3. Push current into the correct spot in tmp_stack
        tmp_stack.append(current)
        
    # At this point, tmp_stack has the largest elements at the top.
    # To restore the original stack structure (with smallest on top), 
    # we copy everything back.
    while len(tmp_stack) > 0:
        stack.append(tmp_stack.pop())
        
    return stack

# --- Example Usage ---
my_stack = [34, 3, 31, 98, 92, 23]
print("Original Stack (top is right):", my_stack)

sort_stack(my_stack)
print("Sorted Stack   (top is right):", my_stack)

File: SortStack.cpp
#include <iostream>
#include <stack>
#include <vector>

void sortStack(std::stack<int>& stack) {
    // Temporary stack to hold elements in sorted order
    std::stack<int> tmpStack;

    while (!stack.empty()) {
        // 1. Pop the top element from the original stack
        int current = stack.top();
        stack.pop();

        // 2. While tmpStack is not empty and its top element is GREATER than current
        while (!tmpStack.empty() && tmpStack.top() > current) {
            // Move elements back to the original stack to clear the way
            stack.push(tmpStack.top());
            tmpStack.pop();
        }

        // 3. Push current into the correct spot in tmpStack
        tmpStack.push(current);
    }

    // Copy elements back so smallest is on top
    while (!tmpStack.empty()) {
        stack.push(tmpStack.top());
        tmpStack.pop();
    }
}

// Helper function to print a stack without destroying the original
void printStack(std::stack<int> s) {
    std::vector<int> elements;
    while (!s.empty()) {
        elements.push_back(s.top());
        s.pop();
    }
    std::cout << "[";
    for (int i = static_cast<int>(elements.size()) - 1; i >= 0; --i) {
        std::cout << elements[i] << (i > 0 ? ", " : "");
    }
    std::cout << "] (top is right)\n";
}

int main() {
    std::stack<int> myStack;
    for (int val : {34, 3, 31, 98, 92, 23}) {
        myStack.push(val);
    }

    std::cout << "Original Stack: ";
    printStack(myStack);

    sortStack(myStack);

    std::cout << "Sorted Stack:   ";
    printStack(myStack);

    return 0;
}